File: app/src/main/java/com/botijasoftware/utils/collision/Capsule.java
package com.botijasoftware.utils.collision;

import com.botijasoftware.utils.Vector3;

public class Capsule {

	private Segment segment;
	public float radius;
	
	public Capsule(Vector3 p0, Vector3 p1, float r) {
		segment = new Segment(p0, p1);
		radius = r;
	}
	
	public void set(Vector3 p0, Vector3 p1) {
		segment.set(p0,p1);
	}
	
	public boolean collides(Vector3 p) {
		return (segment.distancesq(p) <= radius*radius);
	}
	
	public boolean collides(Sphere s) {
		return (segment.distancesq(s.getCenter()) <= (s.radius*s.radius + radius*radius) );
	}
	
	public boolean collides(Ray r) {
		return false;
	}
	
	public boolean collides(Capsule c) {
		return false;
	}
	
	public boolean collides(Segment s) {
		return false;
	}

	public float getRadius() {
		return radius;
	}
	
	public void setRadius(float r) {
		this.radius = r;
	}

	public Vector3 getP0() {
		return segment.p0;
	}
	
	public Vector3 getP1() {
		return segment.p1;
	}
	
}

File: app/src/main/java/com/botijasoftware/utils/collision/Capsule.kt
package com.botijasoftware.utils.collision

import com.botijasoftware.utils.Vector3

class Capsule(p0: Vector3, p1: Vector3, var radius: Float) {

    private val segment: Segment

    init {
        segment = Segment(p0, p1)
    }

    operator fun set(p0: Vector3, p1: Vector3) {
        segment.set(p0, p1)
    }

    fun collides(p: Vector3): Boolean {
        return segment.distancesq(p) <= radius * radius
    }

    fun collides(s: Sphere): Boolean {
        return segment.distancesq(s.center) <= s.radius * s.radius + radius * radius
    }

    fun collides(r: Ray): Boolean {
        return false
    }

    fun collides(c: Capsule): Boolean {
        return false
    }

    fun collides(s: Segment): Boolean {
        return false
    }

    val p0: Vector3
        get() = segment.p0

    val p1: Vector3
        get() = segment.p1

}